[ConstraintSystem] Record originator constraint for each type variable binding

xedin opened the pull request:
This helps with a couple of things: 

- Simplify `PotentialBinding` interface and its memory footprint;

- Provide more contextual information when a binding gets 
   attempted because it's possible to use locator of the binding
  as a source of assignment.

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[ConstraintSystem] Use originating constraint as a source for a binding

Unify all of the fields of `PotentialBinding` which have to do with
location information into a single field with is either a constraint
(for regular bindings) or constraint locator (for "holes").

This helps to reduce the amount of space used by `PotentialBinding`
 as well as propagate more contextual information when binding gets
 attempted.

xedin committed:
[TypeChecker] Don't try to join to function type with different escapiness

Solver needs to attempt both escaping and non-escaping variants
to determine correct error location.

xedin committed:
[CSFix] Convert missing explicit @escaping fix to be a contextual mismatch

Sometimes diagnostic needs both sides of the conversion e.g.
when there is an attempt to bind generic argument to non-escaping type.

xedin committed:
[Diagnostics] Preserve raw from/to types in contextual mismatch diagnostic

This is useful in some situations where access to type variable(s)
helps to diagnose the problem properly e.g. if it's a conversion
to generic parameter.

xedin committed:
[Diagnostics] Convert missing @escaping diagnostic to contextual failure

This gives diagnostic access to both sides of a conversion/binding
which makes it easier to diagnose errors associated with generic parameters.

xedin committed:
[ConstraintSystem] Use binding as a source of type variable assignment

Since `binding` has all of the required information now it's possible
to use its `locator` as a source of type variable assignment
(`Bind` constraint) in `TypeVariableBinding::attempt` which helps
to improve diagnostics.

xedin commented:
/cc @hborla @LucianoPAlmeida

xedin commented:
@swift-ci please smoke test

xedin commented:
@swift-ci please smoke test

xedin commented:
@swift-ci please smoke test

LucianoPAlmeida commented on lib/Sema/ConstraintSystem.h:
```diff
@@ -3288,28 +3288,72 @@ class ConstraintSystem {
+    bool isDefaultableBinding() const {
+      if (auto *constraint = BindingSource.dyn_cast<Constraint *>())
+        return constraint->getKind() == ConstraintKind::Defaultable;
+      // If binding source is not constraint - it's a hole, which is
```
So a "hole" is a type variable that couldn't have a fixed inferred by a constraint in the system? I'm just asking because I don't fully understand what "hole" is :)

xedin replied:
Yeah, a hole is a type variable which we couldn't infer a type for e.g. `_ = .foo` here there is no way to infer implicit base type (represented by `$T_base.Type`) for example, or something like this:

```
func foo<T>() -> T {}
_ = foo()
```

where there is no binding for generic parameter `T`.

LucianoPAlmeida replied:
Humm got it. Thanks :)

xedin replied:
No problem!

theblixguy replied:
Is it related to typed holes in Haskell? Just curious to learn a bit more about how holes work

xedin replied:
Yes, that's pretty one 1-to-1 comparison.

xedin replied:
The only difference is that here this mechanism is only used for diagnostics at the moment but it could be extended to cover e.g. code completion as well.

hborla replied:
They're the same in the sense that they're placeholders for something that is yet to be specified (by the programmer). The difference between typed holes is Haskell and "holes" here is that in Haskell, a hole is a placeholder for another expression. Here, it's a placeholder for a type (that we couldn't infer)

xedin replied:
I think in Swift case as well it would be appropriate to consider holes as placeholders for missing expressions in cases like contextual base or missing argument(s) etc.

hborla replied:
What I mean is that in Haskell, a typed hole is something the programmer writes explicitly in source code in an expression context

xedin replied:
Ah yes, that’s definitely the difference here (for now) :)

LucianoPAlmeida commented on lib/Sema/ConstraintSystem.h:
```diff
@@ -3288,28 +3288,72 @@ class ConstraintSystem {
+      return ConstraintKind::Bind;
+    }
+
+    ConstraintLocator *getLocator() const {
```
Just a question is that since we have a `getSourceKind()`, would this expected to be a `getSourceLocator()`?

xedin replied:
I think it's okay to name this locator because it really belongs to the binding, we need `getSourceKind()` to distinguish between kind of the binding binding accepts (represented by `Kind)` and kind of its source, which I'm not 100% certain is even necessary but I didn't want to extend the scope of changes here.

xedin replied:
It looks like we only use `getSourceKind` in one place and since we've got a locator, could probably use that instead there... Let me give it a try.

xedin replied:
`getSourceKind` has been removed :)

LucianoPAlmeida commented on lib/Sema/CSSimplify.cpp:
```diff
@@ -1548,8 +1548,8 @@ ConstraintSystem::matchFunctionTypes(FunctionType *func1, FunctionType *func2,
 
-    auto *fix = MarkExplicitlyEscaping::create(
-        *this, getConstraintLocator(locator), func2);
+    auto *fix = MarkExplicitlyEscaping::create(*this, func1, func2,
```
Nice!
That's exactly the diagnostic which was failing when I try to use the source locator as the binding locator, I definitely was missing something 😅

xedin committed:
[ConstraintSystem] Remove `getSourceKind()` from `PotentialBinding`

There was a single place where it was used to determine whether
a binding comes from an argument conversion constraint. Since
constraint locator is part of the binding now it's possible to
use it instead and remove unused accessor.

xedin commented:
@swift-ci please smoke test

xedin commented:
@swift-ci please smoke test
